tower-http/src/decompression/request/future.rs

```diff
@@ -0,0 +1,100 @@
+use crate::compression_utils::AcceptEncoding;
+use crate::BoxError;
+use bytes::Buf;
+use http::{header, HeaderValue, Response, StatusCode};
+use http_body::{combinators::UnsyncBoxBody, Body, Empty};
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+pin_project! {
+    #[derive(Debug)]
+    /// Response future of [`RequestDecompression`]
+    ///
+    /// Can either poll its inner future or creates a `Response`
+    /// with an Unsupported Media Type status code.
+    ///
+    /// [`RequestDecompression`]: super::RequestDecompression
+    pub struct RequestDecompressionFuture<F, B, E>
+    where
+        F: Future<Output = Result<Response<B>, E>>,
+        B: Body
+    {
+        #[pin]
+        kind: Kind<F, B, E>,
+    }
+}
+
+pin_project! {
+    #[derive(Debug)]
+    #[project = StateProj]
+    enum Kind<F, B, E>
+    where
+        F: Future<Output = Result<Response<B>, E>>,
+        B: Body
+    {
+        Inner {
+            #[pin]
+            fut: F
+        },
+        Unsupported {
+            #[pin]
+            accept: AcceptEncoding
+        },
+    }
+}
+
+impl<F, B, E> RequestDecompressionFuture<F, B, E>
+where
+    F: Future<Output = Result<Response<B>, E>>,
+    B: Body,
+{
+    #[must_use]
+    pub(super) fn unsupported_encoding(accept: AcceptEncoding) -> Self {
+        Self {
+            kind: Kind::Unsupported { accept },
+        }
+    }
+
+    #[must_use]
+    pub(super) fn inner(fut: F) -> Self {
+        Self {
+            kind: Kind::Inner { fut },
+        }
+    }
+}
+
+impl<F, B, E> Future for RequestDecompressionFuture<F, B, E>
+where
+    F: Future<Output = Result<Response<B>, E>>,
+    B: Body + Send + 'static,
+    <B as Body>::Data: Buf + 'static,
+    <B as Body>::Error: Into<BoxError> + 'static,
```

Comment:
This should work too.

```suggestion
    B::Data: Buf + 'static,
    B::Error: Into<BoxError> + 'static,
```

Reply:
Fixed, I think I needed it when I used a more restrictive signature